File: Source/Aladdin.CAPI.GOST/CAPI/GOST/Keyx/GOSTR3412/KExp15KeyUnwrap.cs
using System;

namespace Aladdin.CAPI.GOST.Keyx.GOSTR3412
{
    ////////////////////////////////////////////////////////////////////////////
    // Согласование ключа KExp15 на стороне-получателе
    ////////////////////////////////////////////////////////////////////////////
    public class KExp15KeyUnwrap : CAPI.TransportKeyUnwrap
    {
        // идентификатор алгоритма согласования ключа
        private ASN1.ISO.AlgorithmIdentifier transportAgreementParameters;     
    
        // конструктор
        public KExp15KeyUnwrap(ASN1.ISO.AlgorithmIdentifier transportAgreementParameters) 
        {   
            // сохранить переданные параметры
            this.transportAgreementParameters = transportAgreementParameters; 
        } 
        public override ISecretKey Unwrap(IPrivateKey privateKey, 
            TransportKeyData transportData, SecretKeyFactory keyFactory)
        {
            // извлечь зашифрованный ключ с параметрами
            ASN1.GOST.GOSTR3410KeyTransport2015 keyTransport = 
                new ASN1.GOST.GOSTR3410KeyTransport2015(
                    ASN1.Encodable.Decode(transportData.EncryptedKey)
            ); 
            // раскодировать открытый ключ
            IPublicKey publicKey = privateKey.Factory.DecodePublicKey(
                keyTransport.EphemeralPublicKey
            ); 
            // создать алгоритм согласования ключа
            using (ITransportAgreement transportAgreement = CreateAgreementAlgorithm(privateKey))
            {
                // расшифровать ключ
                return transportAgreement.Unwrap(privateKey, publicKey, 
                    keyTransport.Ukm.Value, keyTransport.EncryptedKey.Value, keyFactory
                ); 
            }
        }
        protected virtual ITransportAgreement CreateAgreementAlgorithm(IPrivateKey privateKey)
        {
            // создать алгоритм SSDH
            using (ITransportAgreement transportAgreement = KExp15Agreement.CreateSSDH(
                privateKey.Factory, privateKey.Scope, transportAgreementParameters))
            {
                // проверить наличие алгоритма
                if (transportAgreement == null) throw new NotSupportedException(); 

                // вернуть алгоритм ESDH
                return new CAPI.Keyx.ESDH(privateKey.Factory, transportAgreement); 
            }
        }
    }
}

File: Source/Aladdin.CAPI.GOST.Java/src/aladdin/capi/gost/keyx/gostr3412/KExp15KeyUnwrap.java
package aladdin.capi.gost.keyx.gostr3412;
import aladdin.asn1.*;
import aladdin.asn1.iso.*;
import aladdin.asn1.gost.*;
import aladdin.capi.*;
import aladdin.capi.keyx.ESDH;
import java.io.*;

////////////////////////////////////////////////////////////////////////////
// Согласование ключа KExp15 на стороне-получателе
////////////////////////////////////////////////////////////////////////////
public class KExp15KeyUnwrap extends aladdin.capi.TransportKeyUnwrap
{
    // идентификатор алгоритма согласования ключа
    private final AlgorithmIdentifier transportAgreementParameters;     
    
    // конструктор
    public KExp15KeyUnwrap(AlgorithmIdentifier transportAgreementParameters) 
    {   
        // сохранить переданные параметры
        this.transportAgreementParameters = transportAgreementParameters; 
    } 
    @Override public ISecretKey unwrap(aladdin.capi.IPrivateKey privateKey, 
        TransportKeyData transportData, SecretKeyFactory keyFactory) throws IOException
    {
        // извлечь зашифрованный ключ с параметрами
        GOSTR3410KeyTransport2015 keyTransport = new GOSTR3410KeyTransport2015(
            Encodable.decode(transportData.encryptedKey)
        ); 
        // раскодировать открытый ключ
        aladdin.capi.IPublicKey publicKey = privateKey.factory().
            decodePublicKey(keyTransport.ephemeralPublicKey()); 
        
        // создать алгоритм согласования ключа
        try (ITransportAgreement transportAgreement = createAgreementAlgorithm(privateKey))
        {
            // расшифровать ключ
            return transportAgreement.unwrap(privateKey, publicKey, 
                keyTransport.ukm().value(), keyTransport.encryptedKey().value(), keyFactory
            ); 
        }
    }
    protected ITransportAgreement createAgreementAlgorithm(
        IPrivateKey privateKey) throws IOException
    {
        // создать алгоритм согласования ключа (SSDH)
        try (ITransportAgreement transportAgreement = KExp15Agreement.createSSDH(
            privateKey.factory(), privateKey.scope(), transportAgreementParameters))
        {
            // проверить поддержку алгоритмов
            if (transportAgreement == null) throw new UnsupportedOperationException(); 
        
            // вернуть алгоритм согласования ключа (ESDH)
            return new ESDH(privateKey.factory(), transportAgreement); 
        }
    }
}
